File: TMessagesProj/src/main/java/com/overspend1/overgram/ui/liquidglass/LiquidGlassEffect.java
/*
 * This is the source code of Overgram for Android.
 *
 * We do not and cannot prevent the use of our code,
 * but be respectful and credit the original author.
 *
 * Copyright @overspend1, 2024
 */

package com.overspend1.overgram.ui.liquidglass;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import android.graphics.RectF;
import android.os.Build;

import org.telegram.messenger.AndroidUtilities;
import org.telegram.messenger.Utilities;

/**
 * Core liquid glass effect implementation
 * Applies blur, tint, and glassmorphism effects to bitmaps
 */
public class LiquidGlassEffect {
    private GlassParameters parameters;
    private Paint blurPaint;
    private Paint tintPaint;
    private Paint noisePaint;
    private Paint borderPaint;

    private Bitmap cachedBlurredBitmap;
    private long lastCacheTime;
    private static final long CACHE_DURATION = 100; // ms

    public LiquidGlassEffect(GlassParameters parameters) {
        this.parameters = parameters;
        initPaints();
    }

    private void initPaints() {
        // Blur paint with color adjustments
        blurPaint = new Paint(Paint.ANTI_ALIAS_FLAG | Paint.FILTER_BITMAP_FLAG);
        blurPaint.setAlpha((int) (parameters.opacity * 255));

        // Apply saturation and brightness
        ColorMatrix colorMatrix = new ColorMatrix();
        adjustColorMatrix(colorMatrix, parameters.saturation, parameters.brightness);
        blurPaint.setColorFilter(new ColorMatrixColorFilter(colorMatrix));

        // Tint paint
        tintPaint = new Paint();
        tintPaint.setColor(parameters.tintColor);
        tintPaint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_ATOP));

        // Noise paint
        noisePaint = new Paint();
        noisePaint.setAlpha((int) (parameters.noiseIntensity * 255));

        // Border paint
        borderPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        borderPaint.setStyle(Paint.Style.STROKE);
        borderPaint.setStrokeWidth(AndroidUtilities.dp(1));
        borderPaint.setColor(0xFFFFFFFF);
        borderPaint.setAlpha((int) (parameters.borderOpacity * 255));
    }

    /**
     * Apply glass effect to a canvas region
     * @param canvas Target canvas
     * @param bounds Region to apply effect
     * @param backgroundBitmap Background image behind the region
     */
    public void apply(Canvas canvas, RectF bounds, Bitmap backgroundBitmap) {
        if (backgroundBitmap == null || backgroundBitmap.isRecycled()) {
            return;
        }

        try {
            // Extract and blur the background region
            Bitmap blurred = getBlurredBitmap(backgroundBitmap, bounds);
            if (blurred != null && !blurred.isRecycled()) {
                // Draw blurred background
                canvas.drawBitmap(blurred, null, bounds, blurPaint);

                // Apply tint overlay
                if (parameters.tintColor != 0) {
                    canvas.drawRect(bounds, tintPaint);
                }

                // Add subtle noise texture (optional)
                if (parameters.noiseIntensity > 0.01f) {
                    drawNoiseTexture(canvas, bounds);
                }

                // Draw border highlight
                if (parameters.borderOpacity > 0.05f) {
                    canvas.drawRoundRect(bounds, AndroidUtilities.dp(12), AndroidUtilities.dp(12), borderPaint);
                }
            }
        } catch (Exception e) {
            // Fallback to simple semi-transparent background
            Paint fallbackPaint = new Paint();
            fallbackPaint.setColor(0x80000000);
            canvas.drawRoundRect(bounds, AndroidUtilities.dp(12), AndroidUtilities.dp(12), fallbackPaint);
        }
    }

    /**
     * Get blurred version of bitmap region (with caching)
     */
    private Bitmap getBlurredBitmap(Bitmap source, RectF bounds) {
        long currentTime = System.currentTimeMillis();

        // Return cached blur if still valid
        if (cachedBlurredBitmap != null &&
            !cachedBlurredBitmap.isRecycled() &&
            (currentTime - lastCacheTime) < CACHE_DURATION) {
            return cachedBlurredBitmap;
        }

        try {
            // Extract region
            int left = Math.max(0, (int) bounds.left);
            int top = Math.max(0, (int) bounds.top);
            int width = Math.min(source.getWidth() - left, (int) bounds.width());
            int height = Math.min(source.getHeight() - top, (int) bounds.height());

            if (width <= 0 || height <= 0) {
                return null;
            }

            Bitmap region = Bitmap.createBitmap(source, left, top, width, height);

            // Apply blur
            Bitmap blurred = applyBlur(region, parameters.blurRadius);

            // Cache result
            if (cachedBlurredBitmap != null && !cachedBlurredBitmap.isRecycled()) {
                cachedBlurredBitmap.recycle();
            }
            cachedBlurredBitmap = blurred;
            lastCacheTime = currentTime;

            // Clean up temporary bitmap
            if (region != blurred) {
                region.recycle();
            }

            return blurred;
        } catch (Exception e) {
            return null;
        }
    }

    /**
     * Apply blur to bitmap using Utilities.blurBitmap (from Telegram)
     */
    private Bitmap applyBlur(Bitmap bitmap, float radius) {
        if (radius < 1f) {
            return bitmap;
        }

        try {
            // Use Telegram's optimized blur implementation
            int blurRadius = (int) Math.min(25, radius);
            return Utilities.blurBitmap(bitmap, blurRadius, 1, bitmap.getWidth(), bitmap.getHeight(), bitmap.getRowBytes());
        } catch (Exception e) {
            return bitmap;
        }
    }

    /**
     * Adjust color matrix for saturation and brightness
     */
    private void adjustColorMatrix(ColorMatrix matrix, float saturation, float brightness) {
        // Saturation
        float sat = saturation;
        float invSat = 1.0f - sat;
        float R = 0.213f * invSat;
        float G = 0.715f * invSat;
        float B = 0.072f * invSat;

        float[] satMatrix = new float[] {
            R + sat, G,       B,       0, 0,
            R,       G + sat, B,       0, 0,
            R,       G,       B + sat, 0, 0,
            0,       0,       0,       1, 0
        };
        matrix.set(satMatrix);

        // Brightness
        float b = brightness * 255;
        ColorMatrix brightnessMatrix = new ColorMatrix(new float[] {
            1, 0, 0, 0, b,
            0, 1, 0, 0, b,
            0, 0, 1, 0, b,
            0, 0, 0, 1, 0
        });
        matrix.postConcat(brightnessMatrix);
    }

    /**
     * Draw subtle noise texture for realistic glass appearance
     */
    private void drawNoiseTexture(Canvas canvas, RectF bounds) {
        // Simple noise using random pixels (can be optimized with a noise bitmap)
        Paint noisePaint = new Paint();
        noisePaint.setAlpha((int) (parameters.noiseIntensity * 50));

        for (int i = 0; i < 100; i++) {
            float x = bounds.left + (float) Math.random() * bounds.width();
            float y = bounds.top + (float) Math.random() * bounds.height();
            noisePaint.setColor(Math.random() > 0.5 ? 0xFFFFFFFF : 0xFF000000);
            canvas.drawCircle(x, y, 0.5f, noisePaint);
        }
    }

    /**
     * Update parameters and refresh paints
     */
    public void setParameters(GlassParameters parameters) {
        this.parameters = parameters;
        initPaints();
        clearCache();
    }

    /**
     * Clear cached blurred bitmap
     */
    public void clearCache() {
        if (cachedBlurredBitmap != null && !cachedBlurredBitmap.isRecycled()) {
            cachedBlurredBitmap.recycle();
            cachedBlurredBitmap = null;
        }
    }

    /**
     * Release resources
     */
    public void recycle() {
        clearCache();
    }
}

File: TMessagesProj/src/main/java/one/overgram/messenger/ui/liquidglass/LiquidGlassEffect.kt
package one.overgram.messenger.ui.liquidglass

import android.content.Context
import android.graphics.*
import android.os.Build
import android.renderscript.*
import androidx.annotation.RequiresApi
import kotlin.math.min
import kotlin.random.Random

/**
 * Main liquid glass effect implementation
 */
class LiquidGlassEffect(
    private val context: Context,
    private var params: GlassParameters = GlassParameters()
) {
    private val cache = mutableMapOf<String, CachedBlur>()
    private var isAnimating = false
    private var lastRenderTime = 0L
    private var framesSinceLastClear = 0

    // RenderScript (API 17-30)
    private var renderScript: RenderScript? = null
    private var blurScript: ScriptIntrinsicBlur? = null

    companion object {
        private const val CACHE_LIFETIME_MS = 5000L
        private const val MAX_CACHE_ENTRIES = 50

        /**
         * Check if device supports DWM/native blur
         */
        @JvmStatic
        fun supportsNativeBlur(): Boolean {
            return Build.VERSION.SDK_INT >= Build.VERSION_CODES.S // Android 12+
        }
    }

    init {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S &&
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            try {
                renderScript = RenderScript.create(context)
                blurScript = ScriptIntrinsicBlur.create(renderScript, Element.U8_4(renderScript))
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    /**
     * Apply glass effect to canvas
     */
    fun apply(canvas: Canvas, bounds: RectF, background: Bitmap?) {
        if (background == null || background.isRecycled) {
            // Fallback to transparent
            val paint = Paint().apply {
                color = Color.argb((params.backgroundOpacity * 255).toInt(), 0, 0, 0)
            }
            canvas.drawRoundRect(bounds, 12f, 12f, paint)
            return
        }

        // Apply blur
        val effectiveRadius = if (isAnimating && params.adaptiveQuality) {
            params.blurRadius * 0.5f
        } else {
            params.blurRadius
        }

        val blurred = applyBlur(background, effectiveRadius)

        // Apply color adjustments
        val adjusted = applyColorAdjustments(blurred)

        // Apply noise
        val final = if (params.useNoise) {
            applyNoise(adjusted)
        } else {
            adjusted
        }

        // Draw blurred background
        val paint = Paint().apply {
            alpha = (params.backgroundOpacity * 255).toInt()
        }
        canvas.save()
        canvas.clipRect(bounds)
        canvas.drawBitmap(final, bounds.left, bounds.top, paint)
        canvas.restore()

        // Apply tint overlay
        if (Color.alpha(params.tintColor) > 0) {
            val tintPaint = Paint().apply {
                color = params.tintColor
            }
            canvas.drawRoundRect(bounds, 12f, 12f, tintPaint)
        }

        // Draw border
        if (params.borderOpacity > 0) {
            applyBorder(canvas, bounds)
        }

        // Cleanup
        if (final != background && !final.isRecycled) {
            final.recycle()
        }
    }

    /**
     * Apply blur with caching
     */
    fun applyWithCache(canvas: Canvas, bounds: RectF, background: Bitmap?, cacheKey: String) {
        val now = System.currentTimeMillis()

        // Check cache
        cache[cacheKey]?.let { cached ->
            if (cached.region == bounds && (now - cached.timestamp) < CACHE_LIFETIME_MS) {
                // Use cached
                val paint = Paint().apply {
                    alpha = (params.backgroundOpacity * 255).toInt()
                }
                canvas.drawBitmap(cached.blurred, bounds.left, bounds.top, paint)

                if (Color.alpha(params.tintColor) > 0) {
                    canvas.drawRoundRect(bounds, 12f, 12f, Paint().apply { color = params.tintColor })
                }
                if (params.borderOpacity > 0) {
                    applyBorder(canvas, bounds)
                }
                return
            }
        }

        // Not cached, render and cache
        if (background != null && !background.isRecycled) {
            val blurred = applyBlur(background, params.blurRadius)
            val adjusted = applyColorAdjustments(blurred)
            val final = if (params.useNoise) applyNoise(adjusted) else adjusted

            // Store in cache
            cache[cacheKey] = CachedBlur(final.copy(final.config, true), bounds, now)

            // Limit cache size
            if (cache.size > MAX_CACHE_ENTRIES) {
                val oldestKey = cache.entries.minByOrNull { it.value.timestamp }?.key
                oldestKey?.let { key ->
                    cache[key]?.blurred?.recycle()
                    cache.remove(key)
                }
            }

            // Draw
            apply(canvas, bounds, background)
        }
    }

    /**
     * Apply blur effect
     */
    private fun applyBlur(source: Bitmap, radius: Float): Bitmap {
        if (radius <= 0f) return source

        val effectiveRadius = min(radius, params.maxBlurRadius.toFloat())

        return when {
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.S -> {
                // Use RenderEffect API (Android 12+)
                applyRenderEffectBlur(source, effectiveRadius)
            }
            renderScript != null && blurScript != null -> {
                // Use RenderScript (Android 8-11)
                applyRenderScriptBlur(source, effectiveRadius)
            }
            else -> {
                // Fallback to software blur
                applyFastBlur(source, effectiveRadius.toInt())
            }
        }
    }

    /**
     * RenderEffect blur (Android 12+)
     */
    @RequiresApi(Build.VERSION_CODES.S)
    private fun applyRenderEffectBlur(source: Bitmap, radius: Float): Bitmap {
        // Note: RenderEffect is typically used on Views, not Bitmaps directly
        // For bitmap blur on Android 12+, still use RenderScript or software
        return if (renderScript != null) {
            applyRenderScriptBlur(source, radius)
        } else {
            applyFastBlur(source, radius.toInt())
        }
    }

    /**
     * RenderScript blur
     */
    private fun applyRenderScriptBlur(source: Bitmap, radius: Float): Bitmap {
        try {
            val rs = renderScript ?: return source
            val script = blurScript ?: return source

            val input = Allocation.createFromBitmap(rs, source)
            val output = Allocation.createTyped(rs, input.type)

            script.setRadius(min(radius, 25f)) // RenderScript max is 25
            script.setInput(input)
            script.forEach(output)

            val result = Bitmap.createBitmap(source.width, source.height, source.config)
            output.copyTo(result)

            input.destroy()
            output.destroy()

            return result
        } catch (e: Exception) {
            e.printStackTrace()
            return source
        }
    }

    /**
     * Fast blur algorithm (fallback)
     */
    private fun applyFastBlur(source: Bitmap, radius: Int): Bitmap {
        if (radius < 1) return source

        val w = source.width
        val h = source.height
        val pix = IntArray(w * h)
        source.getPixels(pix, 0, w, 0, 0, w, h)

        val wm = w - 1
        val hm = h - 1
        val wh = w * h
        val div = radius + radius + 1

        val r = IntArray(wh)
        val g = IntArray(wh)
        val b = IntArray(wh)

        var rsum: Int
        var gsum: Int
        var bsum: Int
        var x: Int
        var y: Int
        var i: Int
        var p: Int
        var yp: Int
        var yi: Int
        var yw: Int

        val vmin = IntArray(maxOf(w, h))

        var divsum = (div + 1) shr 1
        divsum *= divsum
        val dv = IntArray(256 * divsum)
        i = 0
        while (i < 256 * divsum) {
            dv[i] = i / divsum
            i++
        }

        yw = 0
        yi = 0

        val stack = Array(div) { IntArray(3) }
        var stackpointer: Int
        var stackstart: Int
        var sir: IntArray
        var rbs: Int
        val r1 = radius + 1
        var routsum: Int
        var goutsum: Int
        var boutsum: Int
        var rinsum: Int
        var ginsum: Int
        var binsum: Int

        y = 0
        while (y < h) {
            bsum = 0
            gsum = bsum
            rsum = gsum
            boutsum = rsum
            goutsum = boutsum
            routsum = goutsum
            binsum = routsum
            ginsum = binsum
            rinsum = ginsum

            i = -radius
            while (i <= radius) {
                p = pix[yi + minOf(wm, maxOf(i, 0))]
                sir = stack[i + radius]
                sir[0] = p and 0xff0000 shr 16
                sir[1] = p and 0x00ff00 shr 8
                sir[2] = p and 0x0000ff
                rbs = r1 - kotlin.math.abs(i)
                rsum += sir[0] * rbs
                gsum += sir[1] * rbs
                bsum += sir[2] * rbs
                if (i > 0) {
                    rinsum += sir[0]
                    ginsum += sir[1]
                    binsum += sir[2]
                } else {
                    routsum += sir[0]
                    goutsum += sir[1]
                    boutsum += sir[2]
                }
                i++
            }
            stackpointer = radius

            x = 0
            while (x < w) {
                r[yi] = dv[rsum]
                g[yi] = dv[gsum]
                b[yi] = dv[bsum]

                rsum -= routsum
                gsum -= goutsum
                bsum -= boutsum

                stackstart = stackpointer - radius + div
                sir = stack[stackstart % div]

                routsum -= sir[0]
                goutsum -= sir[1]
                boutsum -= sir[2]

                if (y == 0) {
                    vmin[x] = minOf(x + radius + 1, wm)
                }
                p = pix[yw + vmin[x]]

                sir[0] = p and 0xff0000 shr 16
                sir[1] = p and 0x00ff00 shr 8
                sir[2] = p and 0x0000ff

                rinsum += sir[0]
                ginsum += sir[1]
                binsum += sir[2]

                rsum += rinsum
                gsum += ginsum
                bsum += binsum

                stackpointer = (stackpointer + 1) % div
                sir = stack[stackpointer % div]

                routsum += sir[0]
                goutsum += sir[1]
                boutsum += sir[2]

                rinsum -= sir[0]
                ginsum -= sir[1]
                binsum -= sir[2]

                yi++
                x++
            }
            yw += w
            y++
        }

        x = 0
        while (x < w) {
            bsum = 0
            gsum = bsum
            rsum = gsum
            boutsum = rsum
            goutsum = boutsum
            routsum = goutsum
            binsum = routsum
            ginsum = binsum
            rinsum = ginsum
            yp = -radius * w
            i = -radius
            while (i <= radius) {
                yi = maxOf(0, yp) + x

                sir = stack[i + radius]

                sir[0] = r[yi]
                sir[1] = g[yi]
                sir[2] = b[yi]

                rbs = r1 - kotlin.math.abs(i)

                rsum += r[yi] * rbs
                gsum += g[yi] * rbs
                bsum += b[yi] * rbs

                if (i > 0) {
                    rinsum += sir[0]
                    ginsum += sir[1]
                    binsum += sir[2]
                } else {
                    routsum += sir[0]
                    goutsum += sir[1]
                    boutsum += sir[2]
                }

                if (i < hm) {
                    yp += w
                }
                i++
            }
            yi = x
            stackpointer = radius
            y = 0
            while (y < h) {
                pix[yi] = -0x1000000 and pix[yi] or (dv[rsum] shl 16) or (dv[gsum] shl 8) or dv[bsum]

                rsum -= routsum
                gsum -= goutsum
                bsum -= boutsum

                stackstart = stackpointer - radius + div
                sir = stack[stackstart % div]

                routsum -= sir[0]
                goutsum -= sir[1]
                boutsum -= sir[2]

                if (x == 0) {
                    vmin[y] = minOf(y + r1, hm) * w
                }
                p = x + vmin[y]

                sir[0] = r[p]
                sir[1] = g[p]
                sir[2] = b[p]

                rinsum += sir[0]
                ginsum += sir[1]
                binsum += sir[2]

                rsum += rinsum
                gsum += ginsum
                bsum += binsum

                stackpointer = (stackpointer + 1) % div
                sir = stack[stackpointer]

                routsum += sir[0]
                goutsum += sir[1]
                boutsum += sir[2]

                rinsum -= sir[0]
                ginsum -= sir[1]
                binsum -= sir[2]

                yi += w
                y++
            }
            x++
        }

        val result = Bitmap.createBitmap(w, h, source.config)
        result.setPixels(pix, 0, w, 0, 0, w, h)
        return result
    }

    /**
     * Apply color adjustments
     */
    private fun applyColorAdjustments(bitmap: Bitmap): Bitmap {
        if (params.saturation == 1.0f && params.brightness == 1.0f) {
            return bitmap
        }

        val result = bitmap.copy(bitmap.config, true)
        val canvas = Canvas(result)
        val paint = Paint()

        val cm = ColorMatrix()
        cm.setSaturation(params.saturation)

        if (params.brightness != 1.0f) {
            val scale = params.brightness
            val translate = (1f - scale) * 255
            val matrix = ColorMatrix(floatArrayOf(
                scale, 0f, 0f, 0f, translate,
                0f, scale, 0f, 0f, translate,
                0f, 0f, scale, 0f, translate,
                0f, 0f, 0f, 1f, 0f
            ))
            cm.postConcat(matrix)
        }

        paint.colorFilter = ColorMatrixColorFilter(cm)
        canvas.drawBitmap(bitmap, 0f, 0f, paint)

        return result
    }

    /**
     * Apply noise texture
     */
    private fun applyNoise(bitmap: Bitmap): Bitmap {
        if (params.noiseStrength <= 0f) return bitmap

        val result = bitmap.copy(bitmap.config, true)
        val strength = (params.noiseStrength * 255).toInt()

        for (y in 0 until result.height) {
            for (x in 0 until result.width) {
                val pixel = result.getPixel(x, y)
                val noise = Random.nextInt(-strength, strength + 1)

                val r = ((pixel shr 16 and 0xFF) + noise).coerceIn(0, 255)
                val g = ((pixel shr 8 and 0xFF) + noise).coerceIn(0, 255)
                val b = ((pixel and 0xFF) + noise).coerceIn(0, 255)
                val a = pixel shr 24 and 0xFF

                result.setPixel(x, y, Color.argb(a, r, g, b))
            }
        }

        return result
    }

    /**
     * Draw border
     */
    private fun applyBorder(canvas: Canvas, bounds: RectF) {
        val paint = Paint().apply {
            style = Paint.Style.STROKE
            strokeWidth = 1f
            color = Color.argb((params.borderOpacity * 255).toInt(), 255, 255, 255)
            isAntiAlias = true
        }
        canvas.drawRoundRect(bounds, 12f, 12f, paint)
    }

    /**
     * Update parameters
     */
    fun setParameters(newParams: GlassParameters) {
        params = newParams
        clearCache()
    }

    fun getParameters() = params

    /**
     * Set animating state
     */
    fun setAnimating(animating: Boolean) {
        isAnimating = animating
    }

    /**
     * Clear cache
     */
    fun clearCache() {
        cache.values.forEach { it.blurred.recycle() }
        cache.clear()
    }

    fun clearCache(key: String) {
        cache[key]?.blurred?.recycle()
        cache.remove(key)
    }

    /**
     * Cleanup
     */
    fun destroy() {
        clearCache()
        blurScript?.destroy()
        renderScript?.destroy()
    }

    /**
     * Cached blur data
     */
    private data class CachedBlur(
        val blurred: Bitmap,
        val region: RectF,
        val timestamp: Long
    )
}
